### PR TITLE
Streaming executor fixes #4

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
The streaming doesn't preserve order by default, so the results comparison will need to use set instead of list.

https://github.com/ray-project/ray/issues/32132

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_bulk_executor.py [hunk 1]
```diff
@@ -127,7 +127,11 @@ def test_new_execution_backend_invocation(ray_start_10_cpus_shared):
     DatasetContext.get_current().new_execution_backend = True
```
Comment (ericl) [inline python/ray/data/tests/test_bulk_executor.py:127]: This test is called `test_bulk_executor` so shall we just force bulk execution mode instead?

Comment (jianoaix) [reply python/ray/data/tests/test_bulk_executor.py:127]: Makes sense!

**Comment by jianoaix** [comment]
Failing tests seem not related.

File: python/ray/data/tests/test_dataset.py [hunk 2]
```diff
@@ -384,12 +384,13 @@ def test_zip_different_num_blocks_split_smallest(
...
-    ds = ds1.zip(ds2)
+    ds = ds1.zip(ds2).fully_executed()
+    num_blocks = ds._plan._snapshot_blocks.executed_num_blocks()
     assert ds.take() == [{str(i): i for i in range(num_cols1 + num_cols2)}] * n
     if should_invert:
-        assert ds.num_blocks() == num_blocks2
+        assert num_blocks == num_blocks2
     else:
-        assert ds.num_blocks() == num_blocks1
+        assert num_blocks == num_blocks1
```
Comment (c21) [inline python/ray/data/tests/test_dataset.py:393]: curious as https://github.com/ray-project/ray/pull/32860 is merged, would the change here is still needed?

Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py:393]: Have you tested that PR with streaming enabled for this test? Also note this is to handle issue introduced in https://github.com/ray-project/ray/issues/32132.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Test failures

**Pull request merged**